docs/architecture/ADR-004-interest-and-reserves.md

```diff
@@ -0,0 +1,257 @@
+
+A governance parameter `ReserveFactor` must be kept which specifies the portion of borrow interest that will be funnelled into reserves. This parameter will either be decided per token, or as a single value for all asset types.
+
+> TODO: Decide on the above.
+
+Both the reserved amount of a given token, and the token:uToken exchange rate, increase when interest is _accrued_, rather then when it is _repaid_.
+
+This timing of reserve increases matches the behavior of the [Compound cToken smart contract](https://github.com/compound-finance/compound-protocol/blob/master/contracts/CToken.sol) we are using as a reference - see around line 410 in function accrueInterest.
+
+## Detailed Design
+
+As noted in [ADR-003](./ADR-003-borrow-assets.md), open borrow positions are stored in the`x/leverage` module with the keys `borrowPrefix | lengthPrefixed(borrowerAddress) | tokenDenom` and values of type `sdk.Int`.
+
+When accruing interest, the borrowed amount (`sdk.Int`) must be increased for each open borrow position. The increase should be calculated as follows...
+
+### Borrow Interest Epoch
+
+> accrued interest = borrowed amount × (interest rate × time elapsed)
+
+A governance parameter `BorrowInterestEpoch` with type `int64` will determine the number blocks to wait between interest calculations. Whenever the number of blocks created since the last interest accrual reaches the desired value, interest accrues on all borrow positions simultaneously during `EndBlock`:
+
+```go
+// EndBlock
+// k is the x/leverage keeper and ctx contains information on the current block
+if ctx.BlockHeight() >= k.GetLastInterestBlock() + k.GetBorrowInterestEpoch() {
+    // unix times (int64 values, measured in seconds)
+    secondsElapsed := ctx.BlockTime.Unix() - k.GetLastInterestTime()
+    // accrue interest to all borrow positions at once
+    err = k.AccrueAllInterest(secondsElapsed)
+    // Store the current block's height and time in keeper
+    k.SetLastInterestBlock(ctx.BlockHeight())
+    k.SetLastInterestTime(ctx.BlockTime())
+}
+```
+
+To support the function above, the `x/leverage` module keeper must store the last unix time and block height at which interest accrued. Both values will be marshaled to binary from `sdk.Int`. Single byte prefixes can be used:
+
+```go
+key = interestLastHeightPrefix // e.g. 0x06
+key = interestLastTimePrefix // e.g. 0x07
+```
+
+### Dynamic Borrow Interest Rates
+
+Borrow interest rates are dynamic. They are calculated using the lending pool's current borrow utilization for each asset type, as well as multiple governance parameters that are decided on a per-token basis. Even the model used for the interest rate (which determines what other governance parameters are required) must itself be governed.
+
+Here are some example initial governance parameters which would be stored for a single asset type:
+```go
```

Comment:
I'm not sure I follow this. What is a rate model and why do we need it? When would we ever support multiple at once?